cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/AbstractServiceRoutingHandler.java

```diff
@@ -0,0 +1,318 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.runtime.monitor;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.io.Closeables;
+import io.cdap.cdap.common.BadRequestException;
+import io.cdap.cdap.common.ServiceException;
+import io.cdap.cdap.common.ServiceUnavailableException;
+import io.cdap.cdap.common.discovery.EndpointStrategy;
+import io.cdap.cdap.common.discovery.RandomEndpointStrategy;
+import io.cdap.cdap.common.discovery.URIScheme;
+import io.cdap.cdap.common.security.HttpsEnabler;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.BodyConsumer;
+import io.cdap.http.BodyProducer;
+import io.cdap.http.HttpResponder;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.ProtocolException;
+import java.net.URI;
+import java.net.URL;
+import java.net.UnknownServiceException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import javax.net.ssl.HttpsURLConnection;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link AbstractServiceRoutingHandler} forwards  requests to CDAP services and
+ * relays the response. It doesn't carry out any auth checks. Auth checks can be
+ * performed by child classes or by the downstream CDAP service.
+ */
+public class AbstractServiceRoutingHandler extends AbstractHttpHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      AbstractServiceRoutingHandler.class);
+
+  private final LoadingCache<String, EndpointStrategy> endpointStrategyLoadingCache;
+
+  /**
+   * Handles requests for CDAP services by forwarding them to the correct
+   * destinations and relaying the response.
+   *
+   * @param discoveryServiceClient Discovery client for getting CDAP service
+   *                               addresses.
+   */
+  public AbstractServiceRoutingHandler(
```

Comment:
Prefer to have `protected` constructor for abstract class.

Reply:
Changed and noted.